force-app/main/releaseGating/classes/EDAReleaseGateProvider.cls

```diff
@@ -0,0 +1,125 @@
+        this.releaseGatesMap = new Map<String, ReleaseGateBase>();
+        EDAWinter22ReleaseGate winter22ReleaseGate = new EDAWinter22ReleaseGate();
+        this.releaseGatesMap.put(winter22ReleaseGate.getReleaseGateName(), winter22ReleaseGate);
+    }
+
+    /**
+     * @description Get a list of release gates
+     * @param  version API version
+     * @return         A list of ReleaseGate models
+     */
+    private List<GetReleaseGateActionResultModel.ReleaseGate> getReleaseGateList(Decimal version) {
+        List<GetReleaseGateActionResultModel.ReleaseGate> releaseGateList = new List<GetReleaseGateActionResultModel.ReleaseGate>();
+        for (ReleaseGateBase releaseGate : this.releaseGatesMap.values()) {
+            releaseGateList.add(releaseGate.getReleaseGate(version));
+        }
+        return releaseGateList;
+    }
+
+    /**
+     * @description Get a list of products with release gates
+     * @param  version Callable API version
+     * @return         A list of products and release gates in json format
+     */
+    public virtual GetReleaseGateActionResultModel getReleaseGates(Decimal version) {
+        List<GetReleaseGateActionResultModel.GetReleaseGateResult> resultList = new List<GetReleaseGateActionResultModel.GetReleaseGateResult>{
+            new GetReleaseGateActionResultModel.GetReleaseGateResult(
+                new GetReleaseGateActionResultModel.ReleaseGateProduct(PRODUCT_NAME, Label.stgEDAAppTitle, ''),
+                this.getReleaseGateList(version)
+            )
+        };
+        return new GetReleaseGateActionResultModel(resultList, true, null);
+    }
+
+    /**
+     * @description Activate a release gate
+     * @param  version     Callable API version
+     * @param  productName Release gate product name
+     * @param  gateName    Release gate name
+     * @return             Activate release gate result in json format
+     */
+    public virtual ActivateReleaseGateActionResultModel activateReleaseGate(
+        Decimal version,
+        String productName,
+        String gateName
+    ) {
+        if (productName != PRODUCT_NAME) {
+            throw new IllegalArgumentException(
+                String.format(Label.stgErrorSettingsProductInvalidResponse, new List<Object>{ productName })
+            );
+        }
+        ReleaseGateBase releaseGate = this.releaseGatesMap.get(gateName);
+        if (releaseGate == null) {
+            throw new IllegalArgumentException(
+                String.format(Label.stgErrorSettingsProductInvalidResponse, new List<Object>{ gateName })
+            );
+        }
+        return releaseGate.activateReleaseGate(version, productName, gateName);
+    }
+
+    /**
+     * @description Call a release gating action
+     * @param  action Release gating action name
+     * @param  args   Additional action parameters
+     * @return        Action result in json format
+     */
+    public Object call(String action, Map<String, Object> args) {
```

Comment:
If I do like this I will have to duplicate these 3 lines of code:
```
        String resultJson = JSON.serialize(result);
        UTIL_Debug.debug('Action finished, result: \r\n' + resultJson);
        return resultJson;
```
If we add one more action we will have to add the above lines of code again
Also there will be several return statements, not sure if this will be easier to read

Reply:
Yes, that's a good point, I originally didn't like much idea of repeating code, but now that I am more used to it, I have 2 parts of me fighting all the time :-) hehe

Reply:
In this case, I would prefer to use the guard closes to show that each area of concern is treated separately. The code being duplicated once is an acceptable outcome to show the distinct actions being taken and it forces us to consider the entire use case and cover the code appropriately.

Reply:
Thanks guys, I've made the change

Reply:
thank you! great job